File: detectors/fixtures/go/positive/reverted_guard_before_revert.go
// Fixture: state BEFORE the revert commit (the guard fn exists here).
// validate_transfer_leaves_status is present — this is the guard that
// was later reverted.

package transfer

import "errors"

// Vault holds a simple balance.
type Vault struct {
	Balance uint64
}

// validateTransferLeavesStatus guards against transfers on exited leaves.
func validateTransferLeavesStatus(leafStatus string) error {
	if leafStatus == "exited" {
		return errors.New("leaf already exited to L1; transfer rejected")
	}
	return nil
}

// Transfer executes a transfer after validating leaf status.
func (v *Vault) Transfer(amount uint64, leafStatus string) error {
	if err := validateTransferLeavesStatus(leafStatus); err != nil {
		return err
	}
	v.Balance += amount
	return nil
}

File: detectors/fixtures/rust/positive/reverted_guard_before_revert.rs
// Fixture: state BEFORE the revert commit (the guard fn exists here).
// Used by test_reverted_guard_mine_rust_go.py to build a synthetic Rust repo.
// The guard fn validate_supply_cap is present in this commit.

pub struct Vault {
    pub balance: u64,
}

impl Vault {
    /// Guard: rejects deposits that would push total supply above the cap.
    pub fn validate_supply_cap(&self, amount: u64, cap: u64) -> Result<(), &'static str> {
        if self.balance.saturating_add(amount) > cap {
            return Err("supply cap exceeded");
        }
        Ok(())
    }

    pub fn deposit(&mut self, amount: u64, cap: u64) -> Result<(), &'static str> {
        self.validate_supply_cap(amount, cap)?;
        self.balance += amount;
        Ok(())
    }
}
